robottelo/report_portal/portal.py

```diff
@@ -55,216 +59,119 @@ def headers(self):
 
-        :param str status: Filter tests of a launch with tests `status`
-        :param str defect_type: Filter tests of a launch with tests `defect_type`
+        :param str launch: Dict of a target launch to fetch test items for
+        :param dict test_args: apply the given filters and their values to the search request
         :returns dict: All filtered tests dict based on params data keyed by test name and test
             properties as value, in format -
             ```{'test_name1':test1_properties_dict, 'test_name2':test2_properties_dict}```
         """
+        params = {
+            'page.size': 50,
+            'page.sort': 'name',
+            'filter.eq.launchId': launch["id"],
+            'filter.ne.type': "SUITE",
+        }
+
+        # parse the test filter parameters and turn them into API filter parameters
+        if test_args is None:
+            test_args = {}
+        if not isinstance(test_args, dict):
+            raise TypeError('test_args is expected to be of type dict')
+        if test_args.get('status'):
+            params['filter.in.status'] = ','.join(test_args['status']).upper()
+        if test_args.get('defect_types'):
+            params['filter.in.issueType'] = ','.join(
+                [ReportPortal.defect_types[t] for t in test_args['defect_types']]
+            )
+        if test_args.get('user'):
+            params['filter.has.attributeKey'] = 'assignee'
+            params['filter.has.attributeValue'] = test_args['user']
+
+        # send HTTP request to RP API, retrieve the paginated results and join them together
         page = 1
-        params = self._test_params(status, defect_type, user)
-        total_pages, data = self._test_requester(params=params, page=page)
-        while page < total_pages:
+        total_pages = 1
+        resp_tests = []
+        while page <= total_pages:
+            logger.debug(page)
+            params['page.page'] = page
+            resp = requests.get(
+                url=f'{self.api_url}/item',
+                headers=self.headers,
+                params=params,
+                verify=False,
+            )
+            resp.raise_for_status()
+            resp_tests.extend(resp.json()['content'])
+            total_pages = resp.json()['page']['totalPages']
             page += 1
-            _, pagedata = self._test_requester(params=params, page=page)
-            data.extend(pagedata)
-        # formatting tests data to return tests data keyed by test name
-        tests_ = {test['name']: test for test in data}
-        return tests_
+
+        # Only select tests matching the supplied paths. This is a workaround for RP API limitation
+        # - unable to combine multiple filters of a same type
+        if test_args.get('paths'):
+            resp_tests = [
+                test
+                for test in resp_tests
+                if any([path for path in test_args['paths'] if path in test['name']])
```

Comment:
You could have used a generator instead of list comprehention, which would be more lean.